src/Storages/MergeTree/BoolMask.h

```diff
@@ -1,37 +1,59 @@
 #pragma once
 
+#include <fmt/format.h>
+
 /// Multiple Boolean values. That is, two Boolean values: can it be true, can it be false.
 struct BoolMask
 {
     bool can_be_true = false;
     bool can_be_false = false;
 
     BoolMask() = default;
-    BoolMask(bool can_be_true_, bool can_be_false_) : can_be_true(can_be_true_), can_be_false(can_be_false_) {}
+    BoolMask(bool can_be_true_, bool can_be_false_) : can_be_true(can_be_true_), can_be_false(can_be_false_) { }
 
-    BoolMask operator &(const BoolMask & m) const
-    {
-        return {can_be_true && m.can_be_true, can_be_false || m.can_be_false};
-    }
-    BoolMask operator |(const BoolMask & m) const
-    {
-        return {can_be_true || m.can_be_true, can_be_false && m.can_be_false};
-    }
-    BoolMask operator !() const
+    BoolMask operator&(const BoolMask & m) const { return {can_be_true && m.can_be_true, can_be_false || m.can_be_false}; }
+    BoolMask operator|(const BoolMask & m) const { return {can_be_true || m.can_be_true, can_be_false && m.can_be_false}; }
+    BoolMask operator!() const { return {can_be_false, can_be_true}; }
+
+    bool operator==(const BoolMask & other) const { return can_be_true == other.can_be_true && can_be_false == other.can_be_false; }
+
+    /// Check if mask is no longer changeable under BoolMask::combine.
+    /// We use this condition to early-exit KeyConditions::checkInRange methods.
+    bool isComplete(const BoolMask & initial_mask) const
```

Comment:
initial_mask is a bit confusing here. What values do we expect from it?

Maybe we can do `this->combine(initial_mask).isComplete()` check on the call side?

Reply:
initial_mask argument is removed